solutions/USACO jan19-icy-perimeter.mdx

```diff
@@ -0,0 +1,93 @@
+---
+id: usaco-895
+source: USACO Silver 2019 January
+title: Icy Perimeter
+author: Tanish Tyagi 
+---
+
+[Official Analysis](http://www.usaco.org/current/data/sol_perimeter_silver_jan19.html)
+
+<LanguageSection>
+
+<CPPSection>
+
+## C++ Implementation
+
+```cpp
+//created by Tanish Tyagi
+#include <bits/stdc++.h>
+
+using namespace std;
+using ll = long long;
+
+const int MAXN = 1005;
+
+char grid[MAXN][MAXN]; // the grid itself
+int n;
+bool visited[MAXN][MAXN]; // keeps track of which nodes have been visited
+ll currA = 0;
+ll currP = 0;
+ll area = INT_MIN;
+ll perimeter = INT_MIN;
+
+void setIO(string name, bool includeout = false) { // name is nonempty for USACO file I/O
+	ios_base::sync_with_stdio(0); cin.tie(0); // see Fast Input & Output
+	// alternatively, cin.tie(0)->sync_with_stdio(0);
+	freopen((name + ".in").c_str(), "r", stdin); // see Input & Output
+	if (includeout) {
+		freopen((name + ".out").c_str(), "w", stdout);
+	}
+}
+
+void floodfill(int r, int c) {
+	if (r < 0 || r >= n || c < 0 || c >= n) {
+		currP++;
+		return; // if outside grid
+	}
+	if (visited[r][c]) return; // already visited this square
+	if (grid[r][c] == '.') {
+		currP++;
+		return;
+	}
+	visited[r][c] = true; // mark current square as visited
+	if (grid[r][c] == '#') {
+		currA++;
+	}
+	// recursively call floodfill for neighboring squares
+	floodfill(r, c + 1);
+	floodfill(r, c - 1);
+	floodfill(r - 1, c);
+	floodfill(r + 1, c);
+}
+
+int main() {
+	setIO("perimeter", true);
+	cin >> n;
+	for (int i = 0; i < n; i++) {
+		for (int j = 0; j < n; j++) {
+			cin >> grid[i][j];
+		}
+	}
+	for (int i = 0; i < n; i++) {
+		for (int j = 0; j < n; j++) {
+			if (!visited[i][j] && grid[i][j] == '#') {
+				floodfill(i, j);
+				if (area == currA) {
+					perimeter = min(perimeter, currP);
+				}
+				else {
+					area = max(area, currA);
+					perimeter = max(perimeter, currP);
```

Comment:
I don't think this is right... consider an input like
```
6
###.#.
###.#.
###.#.
....#.
....#.
....#.
```

Reply:
This hasn't been fixed yet...

Reply:
just fixed it, thanks for pointing it out!